ta/pkcs11/src/processing_symm.c

```diff
@@ -149,34 +222,171 @@ allocate_tee_operation(struct pkcs11_session *session,
+					 void **ctx,
+					 size_t *object_size)
+{
+	uint32_t algo = 0;
+	void *hash_ptr = NULL;
+	uint32_t hash_size = 0;
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+
+	/* Find the hash algorithm */
+	rc = hmac_to_tee_hash(&algo, mech_id);
+	if (rc)
+		return rc;
+
+	hash_size = TEE_ALG_GET_DIGEST_SIZE(algo);
+	hash_ptr = TEE_Malloc(hash_size, 0);
+	if (!hash_ptr)
+		return PKCS11_CKR_DEVICE_MEMORY;
+
+	rc = pkcs2tee_load_hashed_attr(tee_attr,
+				       TEE_ATTR_SECRET_VALUE,
+				       obj,
+				       PKCS11_CKA_VALUE,
+				       algo,
+				       hash_ptr,
+				       &hash_size);
+	if (rc) {
+		EMSG("No secret/hash error");
+		TEE_Free(hash_ptr);
+		return PKCS11_CKR_FUNCTION_FAILED;
+	}
+
+	/* Store the hashed key in extra_ctx pointer */
+	*ctx = hash_ptr;
+
+	*object_size = hash_size * 8;
+
+	return rc;
+}
+
 static enum pkcs11_rc load_tee_key(struct pkcs11_session *session,
-				   struct pkcs11_object *obj)
+				   struct pkcs11_object *obj,
+				   struct pkcs11_attribute_head *proc_params)
 {
 	TEE_Attribute tee_attr = { };
 	size_t object_size = 0;
 	uint32_t tee_key_type = 0;
+	enum pkcs11_key_type key_type = 0;
 	enum pkcs11_rc rc = PKCS11_CKR_OK;
 	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t max_key_size = 0;
+	uint32_t min_key_size = 0;
 
 	if (obj->key_handle != TEE_HANDLE_NULL) {
 		/* Key was already loaded and fits current need */
 		goto key_ready;
 	}
 
-	if (!pkcs2tee_load_attr(&tee_attr, TEE_ATTR_SECRET_VALUE,
-				obj, PKCS11_CKA_VALUE)) {
-		EMSG("No secret found");
-		return PKCS11_CKR_FUNCTION_FAILED;
-	}
-
-	rc = pkcs2tee_key_type(&tee_key_type, obj);
-	if (rc)
-		return rc;
-
 	object_size = get_object_key_bit_size(obj);
 	if (!object_size)
 		return PKCS11_CKR_GENERAL_ERROR;
 
+	switch (proc_params->id) {
+	case PKCS11_CKM_MD5_HMAC:
+	case PKCS11_CKM_SHA_1_HMAC:
+	case PKCS11_CKM_SHA224_HMAC:
+	case PKCS11_CKM_SHA256_HMAC:
+	case PKCS11_CKM_SHA384_HMAC:
+	case PKCS11_CKM_SHA512_HMAC:
+		key_type = get_key_type(obj->attributes);
+		/*
+		 * If Object Key type is PKCS11_CKK_GENERIC_SECRET,
+		 * determine the tee_key_type using the
+		 * mechanism instead of object key_type.
+		 */
+		if (key_type == PKCS11_CKK_GENERIC_SECRET)
+			rc = pkcsmech2tee_key_type(&tee_key_type,
+						   proc_params->id);
+		else
+			rc = pkcs2tee_key_type(&tee_key_type, obj);
+
+		if (rc)
+			return rc;
+
+		mechanism_supported_key_sizes(proc_params->id,
+					      &min_key_size,
+					      &max_key_size);
+
+		if ((object_size / 8) > max_key_size) {
+			rc = hash_secret_helper(proc_params->id,
+						obj,
+						&tee_attr,
+						&session->processing->extra_ctx,
+						&object_size);
+			if (rc)
+				return rc;
+		} else {
+			if (!pkcs2tee_load_attr(&tee_attr,
+						TEE_ATTR_SECRET_VALUE,
+						obj,
+						PKCS11_CKA_VALUE)) {
+				EMSG("No secret found");
+				return PKCS11_CKR_FUNCTION_FAILED;
+			}
+		}
+		break;
+
+	default:
+		break;
+	}
+
+	/*
+	 * For all mechanisms except HMAC, fall over thiscode.
```

Comment:
I would not refer to HMAC here, this inline comment will likely be more generic in the (near?) future.
I even think the instructions block below could move to the `default:` case above, but not a strong opinion.

Reply:
It can be done either ways. I will move it to default.